File: kotlin/src/main/kotlin/com/github/jntakpe/aoc2022/days/Day6.kt
package com.github.jntakpe.aoc2022.days

import com.github.jntakpe.aoc2022.shared.Day
import com.github.jntakpe.aoc2022.shared.readInput

object Day6 : Day {

    override val input = readInput(6)

    override fun part1() = indexOfDifferentChars(input, 4)

    override fun part2() = indexOfDifferentChars(input, 14)

    fun indexOfDifferentChars(input: String, window: Int): Int {
        return input.windowed(window).indexOfFirst { it.toSet().size == it.length } + window
    }
}

File: rust/src/bin/day6.rs
extern crate core;

use std::collections::HashSet;

use itertools::Itertools;

use aoc_2022::shared::{read_input_lines_preserving_empty, Day};

fn main() {
    Day6 {
        input: read_input_lines_preserving_empty(6)
            .first()
            .unwrap()
            .to_string(),
    }
    .run()
}

struct Day6 {
    input: String,
}

impl Day<usize> for Day6 {
    fn part1(&self) -> usize {
        self.index_of_marker(4)
    }

    fn part2(&self) -> usize {
        self.index_of_marker(14)
    }
}

impl Day6 {
    fn index_of_marker(&self, window: usize) -> usize {
        self.input
            .as_bytes()
            .windows(window)
            .map(|a| HashSet::<u8>::from_iter(a.iter().cloned()).len())
            .enumerate()
            .find_or_first(|(_, s)| *s == window)
            .map(|(i, _)| i + window)
            .unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part1_sample() {
        assert_eq!(sample_day().part1(), 11)
    }

    #[test]
    fn part2_sample() {
        assert_eq!(sample_day().part2(), 26)
    }

    fn sample_day() -> Day6 {
        Day6 {
            input: vec!["zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw"]
                .iter()
                .map(|s| s.to_string())
                .collect(),
        }
    }
}
